### PR TITLE
7.x islandora 1872 - Allow custom value separator for Solr

### PR DESCRIPTION
**Pull request by BrandonMorr**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-1872

# What does this Pull Request do?

This will allow the user to specify a separator between multivalued fields in solr metadata display.

# What's new?

* added new textfield in the general configuration form 'islandora_solr_metadata_field_value_separator'.
* added a check for empty values in the admin form validation handler. This is to prevent users from saving the form and having empty values passed to the display, causing results to look messy.
* the separator usage has been added to islandora-solr-metadata-description.tpl.php & islandora-solr-metadata-display.tpl.php templates.

# How should this be tested?

* Ingest an object and create some multivalued fields (mods_name_personal_creator_namePart_ms for example).
* Navigate to your.site/admin/islandora/search/islandora_solr/metadata and under the 'General Configuration' tab you will see a new field labelled 'Field value separator'.
* Enter test values, and also test that empty values are being set to NULL and falling back to default separators.
* Make sure your multivalued field is in a solr metadata display configuration.

# Interested parties
@Islandora/7-x-1-x-committers, @nhart

### CODE REVIEW

File: includes/general.admin.inc [hunk 1]
```diff
@@ -24,6 +24,12 @@ function islandora_solr_metadata_general_admin_form($form, &$form_state) {
       '#description' => t('Only show unique values from each field. NOTE: Uniqueness is checked on values directly out of Solr. Later formatting of the value could break uniqueness.'),
       '#default_value' => variable_get('islandora_solr_metadata_dedup_values', FALSE),
     ),
+    'islandora_solr_metadata_field_value_separator' => array(
+      '#type' => 'textfield',
+      '#title' => t('Field value separator'),
+      '#description' => t('Characters to separate values in multivalued fields.'),
+      '#default_value' => variable_get('islandora_solr_metadata_field_value_separator', ''),
```
Comment (jordandukart) [inline includes/general.admin.inc]: This default isn't the same default being used in the templates.

File: theme/islandora-solr-metadata-description.tpl.php [hunk 2]
```diff
@@ -21,12 +21,12 @@
         print ($desc_array['display_label']); ?>
         <?php endif; ?></h2>
       <?php foreach($description as $value): ?>
-        <p property="description"><?php print check_markup(implode("\n", $value['value']), 'islandora_solr_metadata_filtered_html'); ?></p>
+        <p property="description"><?php print check_markup(implode(variable_get('islandora_solr_metadata_field_value_separator', "\n"), $value['value']), 'islandora_solr_metadata_filtered_html'); ?></p>
```
Comment (jordandukart) [inline theme/islandora-solr-metadata-description.tpl.php]: Should probably set the separator as a variable in the preprocess and then use that value here as opposed to `variable_get`ing.

File: theme/theme.inc [hunk 3]
```diff
@@ -138,11 +138,12 @@ function template_process_islandora_solr_metadata_description(array &$variables)
         if ($field['description_data']['truncation']['truncation_type'] == 'whole_field_option') {
           $updated_value[] = islandora_solr_truncate_field_display(
             $field['value'],
+            "Brandon",
```
Comment (jordandukart) [inline theme/theme.inc]: Debug?

File: theme/theme.inc [hunk 4]
```diff
@@ -154,7 +155,7 @@ function template_process_islandora_solr_metadata_description(array &$variables)
               $field['description_data']['truncation']['ellipsis'],
               $field['description_data']['truncation']['word_safe'],
               $field['description_data']['truncation']['min_wordsafe_length'],
-              " "
+              variable_get('islandora_solr_metadata_field_value_separator', ' ')
```
Comment (jordandukart) [inline theme/theme.inc]: This is different than the default that's being set?

**Review by jordandukart — CHANGES_REQUESTED** [review]

File: includes/general.admin.inc [hunk 5]
```diff
@@ -37,12 +43,22 @@ function islandora_solr_metadata_general_admin_form($form, &$form_state) {
 }
 
 /**
+ * Validation handler for general config form.
+ */
+function islandora_solr_metadata_general_admin_form_validate(&$form, &$form_state) {
```
Comment (jordandukart) [inline includes/general.admin.inc]: If this validation function exists to make sure there is a separator value entered why not use `#required` on the element itself?

File: includes/general.admin.inc (hunk 1 above)
Comment (BrandonMorr) [reply includes/general.admin.inc]: I was wrong. Needs more work.

File: theme/theme.inc [hunk 6]
```diff
@@ -154,7 +157,7 @@ function template_process_islandora_solr_metadata_description(array &$variables)
               $field['description_data']['truncation']['ellipsis'],
               $field['description_data']['truncation']['word_safe'],
               $field['description_data']['truncation']['min_wordsafe_length'],
-              " "
```
Comment (jordandukart) [inline theme/theme.inc:157]: Replacing a " " with a newline is a regression. I'd be fine with hard coding the space for the time being seeing as it only applies in a very specific edge case where the description is being truncated and I don't think warrants having another variable for just this specific option.

File: includes/general.admin.inc [hunk 7]
```diff
@@ -24,6 +24,13 @@ function islandora_solr_metadata_general_admin_form($form, &$form_state) {
       '#description' => t('Only show unique values from each field. NOTE: Uniqueness is checked on values directly out of Solr. Later formatting of the value could break uniqueness.'),
       '#default_value' => variable_get('islandora_solr_metadata_dedup_values', FALSE),
     ),
+    'islandora_solr_metadata_field_value_separator' => array(
+      '#type' => 'textfield',
+      '#title' => t('Field value separator'),
+      '#description' => t('Characters to separate values in multivalued fields.'),
+      '#default_value' => variable_get('islandora_solr_metadata_field_value_separator', "\n"),
```
Comment (qadan) [inline includes/general.admin.inc:31]: I don't think this whole \n business is really gonna work ...

- It's a programming convention and not really UI-friendly
- It's not gonna show up in the config
- It can't easily be entered in the config (typing \n in there doesn't really give you what you want either)
- Saving the config with it in there as the default value is considered 'invalid' and makes it impossible to save the form, even though what's in there is technically valid

A possible compromise here would be to not have the field be required, and allow it to be left blank to use newlines as separators.

Comment (nhart) [reply includes/general.admin.inc:31]: @qadan which is what this commit was doing if I understand what you're saying

https://github.com/Islandora/islandora_solr_metadata/pull/51/commits/fb2a31be90e33bc359e9dcd0770fd53ae7ad492c

Comment (qadan) [reply includes/general.admin.inc:31]: Yep, precisely so. I think clearing out the validator per @jordandukart's suggestion, getting rid of `#required`, and re-adding the \n replacement (and documenting it in the field description) would set this right

**Comment by BrandonMorr** [comment]
@whikloj this has passed internal review and testing, could you take a look? Thanks!

**Review by jordandukart — APPROVED** [review]

**Comment by nhart** [comment]
@whikloj are you able to have a look at this one as well? same functionality as https://github.com/Islandora/islandora_solr_search/pull/302, just in Solr metadata

**Comment by whikloj** [comment]
Yeah I can do that, just working on @adam-vessey's right now.

**Comment by nhart** [comment]
Thank you @whikloj!   I appreciate the time you've put into testing this feature for metadata and search.

**Comment by BrandonMorr** [comment]
Thanks @whikloj!
